Feature implement train test split

ghoersti opened the pull request:
This PR adds functionality for a train test split to letsql. 

 I have not updated the crate to bump semantic version.

```python
  import letsql as ls 
  table = memtable([(i, "val") for i in range(N)], columns=["key1", "val"])
  train_table, test_table = ls.train_test_split(
      table, unique_key="key1", test_sizes=test_size, random_seed=42
  )
```

There was a name space conflict of some sort so there is some odd imports at the top `api.py` , this will be fixed. 
It needs more tests on errors & maybe a warning about how bin size & its impact on  train test split percentages, but. I wanted to get something out there to receive feedback. 

if `test_size` is a `float` you will get back   train , test sets with a proportion of (1 - test_size) , ( test_size) 
Otherwise `test_sizes` must be an iterable of floats that sum to 1.

```python
# If test sizes is a list of floats , mutually exclusive deferred partitions will be returned
partition_proportions = [0.1, 0.2, 0.3, 0.4]


hold_out, test, validation, training = ls.train_test_splits(
    table,
    unique_key="key1",
    test_sizes=partition_proportions,
    num_buckets=N,
    random_seed=42,
)
```

ghoersti commented:
In order to follow the the `CONTRIBUTING.MD` I needed to add these packages to the python env. 

`duckdb pins gcsfs datafusion`  along with having the appraise env vars populated for Postgres.

mesejo commented:
> In order to follow the the `CONTRIBUTING.MD` I needed to add these packages to the python env.
> 
> `duckdb pins gcsfs datafusion` along with having the appraise env vars populated for Postgres.

I'll will update the `CONTRIBUTING.MD`, thanks for pointing it out.

ghoersti commented on python/letsql/expr/api.py:
```diff
@@ -1655,3 +1659,96 @@ def to_parquet(expr: ir.Expr, path: str | Path, **kwargs: Any):
+    train_flag = f"train_{random_str}"
+    comb_key = ibis.literal(",").join(table[col].cast("str") for col in unique_key)
+
+    table = table.mutate(
```
That just makes a combined key from either the string or the list provided. 
The whole thing is stochastic. 

The way that it tries to get the % is this condition
```python
(comb_key + random_str).hash().abs() % num_buckets
            < int((1 - test_size) * num_buckets)
```

The whole thing is stochastic. 
What happens is that `(comb_key + random_str).hash().abs() ` becomes a positive int ( there are some collision risks here)  and its relying on a uniform hash distribution , for small numbers it isn't.  So the default buckets is `100` 

So it is basically just saying hey turn the hash into an integer between 0-99  if the hash happens to be be 
`hash <  .75 * 99 `  its in the train set.

ghoersti replied:
@mesejo  I updated comment for clarity

dlovell replied:
If we are going to use buckets, I think we should use a larger number. Perhaps at least 10,000.

ghoersti replied:
With the default being 10k the data will need to large other wise you run the risk of everything falling in the train set set.  If your data set is less than the num buckets we should probably have a warning.

ghoersti replied:
if  default is 10k for `num_buckets` , if someone has a 2k rows. Every thing could easily end up in train set

dlovell commented on python/letsql/expr/api.py:
```diff
@@ -1655,3 +1659,96 @@ def to_parquet(expr: ir.Expr, path: str | Path, **kwargs: Any):
+        rndm_seed(random_seed)
+
+    # Generate a random 256-bit key
+    random_str = str(grb(256))
```
I think this is better expressed as
`str(random.Random(random_seed).getrandbits(256))`
so that it cannot be impacted by anything outside

ghoersti replied:
I think we should sure that the seed can still use system entropy? 


```python
  # Set the random seed, if needed  for reproducibility
  if random_seed:
      random_str = str(random.Random(random_seed).getrandbits(256))
  else:
      #use system entropy to get seed os.urandom() 
      random_str = str(random.Random().getrandbits(256))

```
Users may want to rely on randomness generated  and not rely on providing a seed because the seed provided may not provide enough entropy to be random. 

I think this  option should be there , vs something where we pick a default seed. 


My preference is that we

ghoersti replied:
also it still has a scoping  issue with 
https://github.com/letsql/letsql/blob/497e33364dcf81b645962cc1c52c20f1c1b78dbb/python/letsql/expr/api.py#L460

ghoersti replied:
we can do this though `__import__("random").Random().getrandbits(1)`

ghoersti replied:
or use `importlib`.

dlovell replied:
when `random_seed` is `None`, `random.Random()` and `random.Random(random_seed)` are the same
how about `from random import Random`?

dlovell replied:
on second thought, i think we should avoid putting implementation into `api.py`. how about we put this somewhere else and expose it through `api.py`?

ghoersti replied:
>when random_seed is None, random.Random() and random.Random(random_seed)
Ahh yeah true .

I can move it somehwere else.

dlovell commented:
@ghoersti How would you feel about extending this so that the user can provide multiple weights to get multiple splits?

ghoersti commented:
> @ghoersti How would you feel about extending this so that the user can provide multiple weights to get multiple splits?

I can do that , but also they could just use a list comprehension and get `[ tuple[ir.Table, ir.Table], tuple[ir.Table, ir.Table]....]`

ghoersti commented:
I implemented a functions `train_test_split` & `train_test_splits` , I also think it should be clear if you change your bucket size your seed no longer reproduces the same sets.

dlovell commented:
> I implemented a functions `train_test_split` & `train_test_splits` and added tests. and removed the function from the api and placed it in `ml.py` I can add tests in `api,py` , I did not add tests for the stubs in `api.py` but I did test the implementations. @dlovell @mesejo

I think that `train_test_split` should be a special case of `train_test_splits` and that `train_test_splits` should take an iterable of `float` that should sum to `<=1` so that the exprs that are returned are mutually exclusive but collectively exhaustive of the rows of the input expr

ghoersti commented:
> > I implemented a functions `train_test_split` & `train_test_splits` and added tests. and removed the function from the api and placed it in `ml.py` I can add tests in `api,py` , I did not add tests for the stubs in `api.py` but I did test the implementations. @dlovell @mesejo
> 
> I think that `train_test_split` should be a special case of `train_test_splits` and that `train_test_splits` should take an iterable of `float` that should sum to `<=1` so that the exprs that are returned are mutually exclusive but collectively exhaustive of the rows of the input expr

Ok this makes sense, I will implement that, should the api still have both exposed ? or just train test splits ?

ghoersti commented:
@dlovell  would we want to change the name to `train_test_partitions` ? Since now this is just a way to partition deferred?

ghoersti commented on python/letsql/expr/ml.py:
```diff
@@ -0,0 +1,148 @@
+
+    if not all(0 < size < 1 for size in test_sizes):
+        raise ValueError("All test sizes must be between 0 and 1.")
+    if not sum(test_sizes) == 1:
```
This does not need a test it is covered in the `train_test_splits`

ghoersti commented on python/letsql/expr/ml.py:
```diff
@@ -0,0 +1,143 @@
+    table = table.mutate(**{"hash": (comb_key + random_str).hash().abs() % num_buckets})
+    train_test_filters = [
+        (
+            literal(bound[0]).cast("decimal(38, 9)") * num_buckets <= table.hash
```
`cast("decimal(38, 9)") ` because default when it executed on duck_db  was DECIMAL(18) and a test was breaking due to overflow.

dlovell commented:
@ghoersti can you implement `visit_Hash` so we can execute via `ls.execute`?

```
File /nix/store/wv660b4m8dydvygpa7cwkrblk041xc7k-letsql-dev-env/lib/python3.10/site-packages/ibis/backends/sql/compilers/base.py:1514, in SQLGlotCompiler.visit_Undefined(self, op, **_)
   1513 def visit_Undefined(self, op, **_):
-> 1514     raise com.OperationNotDefinedError(
   1515         f"Compilation rule for {type(op).__name__!r} operation is not defined"
   1516     )

OperationNotDefinedError: Compilation rule for 'Hash' operation is not defined
```

ghoersti commented:
> @ghoersti can you implement `visit_Hash` so we can execute via `ls.execute`?
> 
> ```
> File /nix/store/wv660b4m8dydvygpa7cwkrblk041xc7k-letsql-dev-env/lib/python3.10/site-packages/ibis/backends/sql/compilers/base.py:1514, in SQLGlotCompiler.visit_Undefined(self, op, **_)
>    1513 def visit_Undefined(self, op, **_):
> -> 1514     raise com.OperationNotDefinedError(
>    1515         f"Compilation rule for {type(op).__name__!r} operation is not defined"
>    1516     )
> 
> OperationNotDefinedError: Compilation rule for 'Hash' operation is not defined
> ```


> @ghoersti can you implement `visit_Hash` so we can execute via `ls.execute`?
> 
> ```
> File /nix/store/wv660b4m8dydvygpa7cwkrblk041xc7k-letsql-dev-env/lib/python3.10/site-packages/ibis/backends/sql/compilers/base.py:1514, in SQLGlotCompiler.visit_Undefined(self, op, **_)
>    1513 def visit_Undefined(self, op, **_):
> -> 1514     raise com.OperationNotDefinedError(
>    1515         f"Compilation rule for {type(op).__name__!r} operation is not defined"
>    1516     )
> 
> OperationNotDefinedError: Compilation rule for 'Hash' operation is not defined
> ```

@dan oh ok  would go in `compiler.py`  for each of our backends that do not nativley support Hash ?

dlovell commented:
> @dan oh ok would go in `compiler.py` for each of our backends that do not nativley support Hash ? Probably something using pyarrow.compute.sha256 to hash.

I don't think we have the choice to use `pyarrow.compute.sha256` (though somehow it might still end up being the equivalent). For our `letsql` backend and `datafusion`, I think it will be something that translates to `digest(..., 'sha256')`.

ghoersti commented:
> > @dan oh ok would go in `compiler.py` for each of our backends that do not nativley support Hash ? Probably something using pyarrow.compute.sha256 to hash.
> 
> I don't think we have the choice to use `pyarrow.compute.sha256` (though somehow it might still end up being the equivalent). For our `letsql` backend and `datafusion`, I think it will be something that translates to `digest(..., 'sha256')`.

I see it looks like `SQLGlotCompiler` already has one: 

```python
>>> [ c for c in dir(SQLGlotCompiler) if 'Hash' in c ] 
['visit_Hash', 'visit_HashBytes']
```

dlovell commented:
@ghoersti can we refactor the example so we don't recalculate counts?

```
name_to_proportion = {                                                                
    "hold_out": 0.1,                                                                  
    "test": 0.2,                                                                      
    "validation": 0.3,        
    "training": 0.4,                                                                                                                                                        
}                                          
partitions = tuple(ls.train_test_splits(   
    table,                                                                            
    unique_key="key1",                                                                
    test_sizes=tuple(name_to_proportion.values()),                                    
    num_buckets=N,                         
    random_seed=42,                                                                   
))                                                                                                                                                                          
name_to_count = {                                                                                                                                                           
    name: partition.count().execute()
    for name, partition in zip(name_to_proportion, partitions)
}                                          
total = sum(name_to_empirical_proportion.values())
for name, count in name_to_count.items():
    print(f"{name} ratio: {count/total}")
```

dlovell commented:
> > > @dan oh ok would go in `compiler.py` for each of our backends that do not nativley support Hash ? Probably something using pyarrow.compute.sha256 to hash.
> > 
> > 
> > I don't think we have the choice to use `pyarrow.compute.sha256` (though somehow it might still end up being the equivalent). For our `letsql` backend and `datafusion`, I think it will be something that translates to `digest(..., 'sha256')`.
> 
> I see it looks like `SQLGlotCompiler` already has one:
> 
> ```python
> >>> [ c for c in dir(SQLGlotCompiler) if 'Hash' in c ] 
> ['visit_Hash', 'visit_HashBytes']
> ```

The `duckdb` backend works, but the `datafusion` backend and our backend don't. If we don't implement `visit_Hash` for them here, can you open an issue for them?

mesejo commented:
> > @dan oh ok would go in `compiler.py` for each of our backends that do not nativley support Hash ? Probably something using pyarrow.compute.sha256 to hash.
> 
> I don't think we have the choice to use `pyarrow.compute.sha256` (though somehow it might still be the equivalent). For our `letsql` backend and `datafusion`, I think it will be something that translates to `digest(..., 'sha256')`.

There is a `visit_HexDigest` in `compiler.py` that can be used as a guide:

```python
    def visit_HexDigest(self, op, *, arg, how):
        return self.f.encode(self.f.digest(arg, how), "hex")
```

So, a possible implementation for hash could be:

```python
    def visit_Hash(self, op, *, arg):
        return self.f.digest(arg, 'sha256')
```

But that returns a binary string, and it is not clear to me that it can be converted into an integer in a performant manner

ghoersti commented:
> name_to_proportion = {                                                                
>     "hold_out": 0.1,                                                                  
>     "test": 0.2,                                                                      
>     "validation": 0.3,        
>     "training": 0.4,                                                                                                                                                        
> }

```python
In [17]: # If test sizes is a list of floats , mutually exclusive partitions will be returned
    ...: partion_names = {
    ...:     "hold_out": 0.1,
    ...:     "test": 0.2,
    ...:     "validation": 0.3,
    ...:     "training": 0.4,
    ...: 
    ...: }
    ...: 
    ...: partitions = tuple(ls.train_test_splits(
    ...:         table,
    ...:         unique_key="key1",
    ...:         test_sizes=list(partion_info.values()),
    ...:         num_buckets=N,
    ...:         random_seed=42,
    ...:     )
    ...: )
    ...: counts = [p.count().execute() for p in  partitions ]
    ...: total = sum(counts)
    ...: 
    ...: for i, partition_name in enumerate(partion_info.keys()):
    ...:     print(f"{partition_name.upper()} Ratio: {round(counts[i]/ total,2)}")
    ...: 
HOLD_OUT Ratio: 0.1
TEST Ratio: 0.2
VALIDATION Ratio: 0.3
TRAINING Ratio: 0.4
```

Done

dlovell commented on python/letsql/expr/api.py:
```diff
@@ -1709,3 +1712,7 @@ def to_parquet(
                 writer.write_batch(batch)
+
+
+def train_test_splits(*args, **kwargs) -> Iterator[tuple[ir.Table, ir.Table]]:
+    return ml.train_test_splits(*args, **kwargs)
```
when we do this, we lose the signature
can we use
`from letsql.expr.ml import train_tests_splits`
instead?

dlovell replied:
@ghoersti why do we redefine `train_test_splits` in `api.py` rather than importing from `ml.py`?

dlovell replied:
this is the impact of redefining with `*args, **kwargs` rather than importing
```
In [1]: import letsql as ls

In [2]: ls.train_test_splits
Out[2]: <function letsql.expr.api.train_test_splits(*args, **kwargs) -> 'Iterator[tuple[ir.Table, ir.Table]]'>

In [3]: ls.expr.ml._train_test_splits
Out[3]: <function letsql.expr.ml._train_test_splits(table: ibis.expr.types.relations.Table, unique_key: str | list[str], test_sizes: Union[Iterable[float], float], num_buckets: int = 10000, random_seed: int | None = None) -> Iterator[ibis.expr.types.relations.Table]>
```

dlovell commented:
All the tests use `memtable`, can we parametrize by `con` and test `duckdb`, `datafusion`, `letsql` and at least one backend we expect to `xfail`?

ghoersti commented:
> All the tests use `memtable`, can we parametrize by `con` and test `duckdb`, `datafusion`, `letsql` and at least one backend we expect to `xfail`?

I can do this.

dlovell commented:
@ghoersti in the other tests, we use connections themselves as test fixtures. when done properly, that makes sure we have "clean" connections so that, for instance, if the postgres test fails in the middle, we aren't running other tests from an unknown state.
i don't know that we need to do that here, and iirc, its a pain to parametrize over the cons, so consider this approach instead
```
diff --git a/python/letsql/tests/test_ml.py b/python/letsql/tests/test_ml.py
index e6e420f..c5ca057 100644
--- a/python/letsql/tests/test_ml.py
+++ b/python/letsql/tests/test_ml.py
@@ -183,15 +183,11 @@ def test_train_test_splits_num_buckets_gt_one():
 
 
 @pytest.mark.parametrize(
-    "connection_name,expected_conn_type",
-    [
-        ("letsql", ls.backends.let.Backend),
-        ("duckdb", ls.backends.duckdb.Backend),
-        ("postgres", ls.backends.postgres.Backend),
-    ],
+    "connect_method",
+    (ls.connect, ls.duckdb.connect, ls.postgres.connect_env),
 )
 def test_train_test_splits_intersections_parameterized_pass(
-    create_passing_connections, connection_name, expected_conn_type
+    connect_method,
 ):
     # This is testing the base case where a single float becomes ( 1-test_size , test_size ) proportion
     # Check counts and overlaps in train and test dataset
@@ -200,9 +196,9 @@ def test_train_test_splits_intersections_parameterized_pass(
 
     # create test table for backend
     test_df = pd.DataFrame([(i, "val") for i in range(N)], columns=["key1", "val"])
-    con = create_passing_connections[connection_name]
-    test_table_name = f"{connection_name}_test_df"
-    con.create_table(test_table_name, test_df)
+    con = connect_method()
+    test_table_name = f"{con.name}_test_df"
+    con.create_table(test_table_name, test_df, temp=con.name=="postgres")
 
     table = con.table(test_table_name)
 
@@ -246,19 +242,16 @@ def test_train_test_splits_intersections_parameterized_pass(
         == 0
     )
 
-    assert isinstance(con, expected_conn_type)
```

dlovell commented:
@ghoersti this is how to mark an individual test in a parametrized list as xfail
```
diff --git a/python/letsql/tests/test_ml.py b/python/letsql/tests/test_ml.py
index e6e420f..204bdbd 100644
--- a/python/letsql/tests/test_ml.py
+++ b/python/letsql/tests/test_ml.py
@@ -183,15 +183,19 @@ def test_train_test_splits_num_buckets_gt_one():
 
 
 @pytest.mark.parametrize(
-    "connection_name,expected_conn_type",
-    [
-        ("letsql", ls.backends.let.Backend),
-        ("duckdb", ls.backends.duckdb.Backend),
-        ("postgres", ls.backends.postgres.Backend),
-    ],
+    "connect_method",
+    (
+        ls.connect,
+        ls.duckdb.connect,
+        ls.postgres.connect_env,
+        pytest.param(
+            ls.datafusion.connect,
+            marks=pytest.mark.xfail(reason="Compilation rule for 'Hash' operation is not define"),
+        ),
+    ),
 )
```
if we do that, we can remove the other test whose sole purpose is to xfail

dlovell commented on python/letsql/expr/ml.py:
```diff
@@ -0,0 +1,143 @@
+
+    comb_key = literal(",").join(table[col].cast("str") for col in unique_key)
+
+    table = table.mutate(**{"hash": (comb_key + random_str).hash().abs() % num_buckets})
```
i think we need a random string component of the hash column name to prevent collisions with user columns

ghoersti replied:
Oh ok this makes sense for sure.

ghoersti commented:
> ls.postgres.connect_env

why deviate for  `ls.postgres.connect_env` 
?

dlovell commented:
> > ls.postgres.connect_env
> 
> why deviate for `ls.postgres.connect_env` ?

We don't need any of the existing tables / state, we simply want to create a table and using a temporary table isolates the other tests from our changes here
